File: file_processing.py
from io import BytesIO
from docx import Document
import pandas as pd
from PyPDF2 import PdfReader
import streamlit as st

def processar_arquivos(uploaded_file):
    """Processa arquivos e retorna texto extraído."""
    try:
        if uploaded_file.name.endswith(".docx"):
            doc = Document(uploaded_file)
            texto = "\n".join([p.text for p in doc.paragraphs])

        elif uploaded_file.name.endswith(".txt"):
            texto = uploaded_file.read().decode("utf-8")

        elif uploaded_file.name.endswith(".pdf"):
            reader = PdfReader(uploaded_file)
            texto = "\n".join([page.extract_text() for page in reader.pages])

        elif uploaded_file.name.endswith(".csv"):
            df = pd.read_csv(uploaded_file)
            texto = df.to_string()

        elif uploaded_file.name.endswith(".xlsx"):
            df = pd.read_excel(uploaded_file)
            texto = df.to_string()

        else:
            st.error("Tipo de arquivo não suportado. Envie .docx, .txt, .pdf, .csv ou .xlsx.")
            return None

        return texto

    except Exception as e:
        st.error(f"Erro ao processar arquivo: {e}")
        return None

def gerar_docx(conteudo, titulo):
    """Função para gerar um arquivo DOCX a partir do texto fornecido."""
    doc = Document()
    doc.add_heading(titulo, level=1)
    doc.add_paragraph(conteudo)

    buffer = BytesIO()
    doc.save(buffer)
    buffer.seek(0)
    return buffer

File: node/fileProcessing.ts
// fileProcessing.ts
import mammoth from "mammoth";
import pdfParse from "pdf-parse";
import XLSX from "xlsx";
import { Document, Packer, Paragraph, HeadingLevel } from "docx";

/**
 * Processa o arquivo enviado e retorna o texto extraído.
 * Suporta: .docx, .txt, .pdf, .csv e .xlsx.
 */
export async function processFile(file: File): Promise<string | null> {
  const fileName = file.name.toLowerCase();
  try {
    if (fileName.endsWith(".docx")) {
      const arrayBuffer = await file.arrayBuffer();
      const result = await mammoth.extractRawText({ buffer: Buffer.from(arrayBuffer) });
      return result.value;
    } else if (fileName.endsWith(".txt")) {
      return await file.text();
    } else if (fileName.endsWith(".pdf")) {
      const arrayBuffer = await file.arrayBuffer();
      const buffer = Buffer.from(arrayBuffer);
      const data = await pdfParse(buffer);
      return data.text;
    } else if (fileName.endsWith(".csv")) {
      return await file.text();
    } else if (fileName.endsWith(".xlsx")) {
      const arrayBuffer = await file.arrayBuffer();
      const workbook = XLSX.read(arrayBuffer, { type: "array" });
      const firstSheetName = workbook.SheetNames[0];
      const worksheet = workbook.Sheets[firstSheetName];
      return XLSX.utils.sheet_to_csv(worksheet);
    } else {
      console.error("Tipo de arquivo não suportado. Envie .docx, .txt, .pdf, .csv ou .xlsx.");
      return null;
    }
  } catch (error) {
    console.error("Erro ao processar arquivo:", error);
    return null;
  }
}

/**
 * Gera um arquivo DOCX a partir do conteúdo e título informados.
 */
export async function generateDocx(content: string, title: string): Promise<Blob> {
  const doc = new Document({
    sections: [
      {
        children: [
          new Paragraph({
            text: title,
            heading: HeadingLevel.HEADING_1,
          }),
          new Paragraph({
            text: content,
          }),
        ],
      },
    ],
  });
  // Chama toBlob sem o segundo parâmetro para satisfazer a tipagem esperada
  const blob = await Packer.toBlob(doc);
  return blob;
}
